initial commit qca port

gnome71 opened the pull request:
This PR is to start a discussion about my work on a port on QCA. I have the problem, that I need Qt5 environment and how to find it on the users machine. I don't want to make a dependency to qt5 in the CONTROL file, because it is a huge download and very time consuming. 
Also on my machine I can just test it under Windows 10 and my Qt5 installation. So for now only msvc and msvc_64 is supported.
I'm open for any hints and critics - I'm still learning CMake.

Thanks for your help till now 
Alex

Commit:
initial commit to qca branch

msftclas commented:
Hi __@gnome71__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

Orvid commented on ports/qca/0001-fix-path-for-vcpkg.patch:
```diff
@@ -0,0 +1,54 @@
+-  "${CMAKE_CURRENT_BINARY_DIR}/lib/cmake/${QCA_CONFIG_NAME_BASE}/${QCA_CONFIG_NAME_BASE}Config.cmake"
+-  INSTALL_DESTINATION ${QCA_LIBRARY_INSTALL_DIR}/cmake/${QCA_CONFIG_NAME_BASE}
++  "${CMAKE_CURRENT_BINARY_DIR}/share/cmake/${QCA_CONFIG_NAME_BASE}/${QCA_CONFIG_NAME_BASE}Config.cmake"
++  INSTALL_DESTINATION ${CMAKE_BINARY_DIR}/share/cmake/${QCA_CONFIG_NAME_BASE}
```
You want the path to be `/share/qca/` rather than `/share/cmake/`.

ras0219-msft commented:
Unfortunately we really need to have this build against our managed version of Qt. This is to enable static linking scenarios where I have multiple components using Qt; if I want to static link `qca` and `qca_other` into my program, I need them to both be built against the same copy of Qt, otherwise, the symbols _from the two different qt versions_ will conflict and I won't be able to link my program.

I totally agree that the current `qt5` is a really big dependency -- we should do more work in the future to try to break it up into smaller pieces that can be modularly built. Also, remember that you only need to build it once and you can share it among all of your projects :). The layout is designed to be completely relocatable, so you CAN zip it up and move it between machines to reduce startup costs as well.

Other than that, your approach here looks pretty solid. I'll test it out once you fix the issue @Orvid picked out above.

Commit:
[qca] fix cmake file location
Added qt5 Build-Depends to CONTROL file.

Commit:
[qca] fix cmake files dir
Added qt5 as Build-Depends to CONTROL file.

gnome71 commented:
I changed the cmake file locations according to the issue @Orvid mentioned in his review. Also I added the build dependency to qt5 in the CONTROL file. So there is no need to search for qt on the users machine, nor has the user to set any env. variables.

Have a nice weekend!
Alex

gnome71 committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into qca

Commit:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into qca

gnome71 committed:
Merge branch 'qca' of github.com:gnome71/vcpkg into qca

Commit:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into qca

Commit:
Merge branch 'qca' of github:gnome71/vcpkg into qca

ras0219-msft commented:
Looks pretty good! There are four paths that catch my attention though:
- certs\
- mkspecs\
- debug\certs\
- debug\mkspecs\

`certs\` has `rootcerts.pem` in it, `mkspecs\` has a `.prf` file. Are these important and are they in the right place? I feel like the `pem` is probably not needed, but I don't know much about `.prf` files.

(Side note: the command `tree /f D:\src\vcpkg\packages\qca_x64-windows` is very nice to get an overview of the final package)

gnome71 committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into qca

gnome71 committed:
formatting portfile

gnome71 committed:
[qca] added PS script to extract local certs from store

gnome71 committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into qca

gnome71 committed:
Merge branch 'qca' of github.com:gnome71/vcpkg into qca

gnome71 committed:
[qca] import local certs

gnome71 committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into qca

gnome71 committed:
Merge branch 'qca' of ssh://github.com/gnome71/vcpkg into qca

gnome71 committed:
[qca] working on certstore

gnome71 committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into qca

gnome71 committed:
Merge branch 'qca' of github.com:gnome71/vcpkg into qca

gnome71 committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into qca

gnome71 commented:
According to:  https://www.openssl.org/docs/faq.html#USER16
it is up to developers or admins to maintain CAs.
Thats what the QCA Maintainers do with the rootcerts.pem file. According to the README the certs are converted from the Mozilla sources. But its quite outdated.
So my plan is to download a recent version maintained by the CURL developers and comment this in the portfile. So that everyone interested can follow the steps taken.

The portfile now builds the files mentioned by  @ras0219-msft in subdirs of PACKAGE/share/qca/ - where they fit better, I think.

gnome71 committed:
[qca] downloading cacerts from curl homepage, because used version is very outdated.

gnome71 committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into qca

gnome71 committed:
Merge branch 'qca' of github.com:gnome71/vcpkg into qca

Orvid commented:
Alternatively, it might be viable to retrieve the cert-data from mozilla's main mirror and generate at build time. It would also mean that it can be versioned and we can bump the package version when it's updated: https://raw.githubusercontent.com/mozilla/gecko-dev/master/security/nss/lib/ckfw/builtins/certdata.txt

An even better solution might be to still do what I mentioned above but have it as a separate port so that it can be versioned independently. If going that route, it would likely be better to version directly off of nss: https://ftp.mozilla.org/pub/security/nss/releases/NSS_3_27_2_RTM/src/

gnome71 committed:
[qca] Added mk-ca-bundle.pl to generate certstore at build time.

gnome71 committed:
[qca] Added mk-ca-bundle.pl to generate certstore at build time.

gnome71 committed:
Merge branch 'qca' of ssh://github.com/gnome71/vcpkg into qca

gnome71 commented:
By now I'm not able to make an extra port for nss because it seems quite complicated for me. It uses the mozilla-build environment and I've not used that before.
By now I'm downloading [https://raw.githubusercontent.com/mozilla/gecko-dev/master/security/nss/lib/ckfw/builtins/certdata.txt](url) (like @Orvid mentioned) and process this file before building the library.

gnome71 committed:
Merge branch 'qca' of github.com:gnome71/vcpkg into qca

gnome71 committed:
Merge branch 'master' into qca

gnome71 committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into qca

gnome71 committed:
Merge branch 'qca' of github.com:gnome71/vcpkg into qca

gnome71 committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into qca

ras0219-msft commented:
Thanks for working through this!

In the future, it would probably be best to separate out the `.pem` into it's own port, or perhaps even remove it entirely. Since the master CA list is somewhat volatile and security sensitive, it's more difficult to handle appropriately compared to normal software -- a user that wants to lock down all their software dependency versions will still probably want to update their CA list.

gnome71 commented:
I would prefer to remove it too. I will contact the library maintainers to clarify that and maybe remove it on the next version bump. 
Thanks for your guidance, it was a great experience for me to participate the first time on such a cool project. Keep up your good work!!
